forbidigo/patterns.go

```diff
@@ -8,10 +8,21 @@ import (
 )
 
 type pattern struct {
-	pattern *regexp.Regexp
-	msg     string
+	features patternFeature
+	pattern  *regexp.Regexp
+	msg      string
 }
 
+// patternFeature defines what text the pattern is for, for example with full
+// packagename instead of import name. This is a set of bits, so more than one
+// feature might be set.
+type patternFeature int
+
+const (
+	patternFeatureSrc     patternFeature = 0
+	patternFeaturePackage patternFeature = 1
```

Comment:
This could be `1<<iota` if go.mod listed >=1.13 as Go language. With 1.12 it causes an error about signed bit shifts.

Reply:
My preference would be just to record on each pattern whether we expect a package name (i.e. `expectsPackageName bool` rather than partition patterns by kind).  It could be I'm not seeing ahead to the end game, but I think that would allow the intent to be expressed a bit more simply.  It does mean that we need to iterate through all the patterns rather than all the features when we do the matching.

Reply:
The purpose becomes clearer when looking at https://github.com/ashanbrown/forbidigo/pull/22 because then there are two features.

That also works, but my gut feeling is that the code for checking whether a pattern needs to be checked will become more complicated: instead of one field, it needs to check n different booleans.

More verbose, but perhaps also more readable because of that. I'll try that to see what it looks like.

Reply:
I think it's okay. I am using your suggestion to refactor and only initialize strings as needed.